### DISCUSSION TITLE
detect dark mode user preferences in Python

### DISCUSSION THREAD

**Discussion by petergaultney** [zauberzeug/nicegui#2178] [Q&A]
### Question

I see that Quasar does this by default (if we call `ui.run(dark_mode=None)`, but if we opt into this automatic detection, it doesn't help much with custom components (e.g. the JSON Editor that is built in, which has its own custom dark mode and a [custom method of activating it](https://github.com/josdejong/svelte-jsoneditor/?tab=readme-ov-file#dark-theme) (`class: jse-theme-dark`).

If I could detect the user's preference here, then it's pretty trivial to conditionally apply `.classes("jse-theme-dark")` to the Element in Python. 

I tried `ui.dark_mode().value` but that actually resets everything to light mode (False) by default and does not give me the access I want. I don't see anything obvious on the `ui` object itself, either.

**Comment by falkoschindler** [comment]
Good question, @petergaultney! You can ask the client using a JavaScript call like this:
```py
async def get_dark_mode():
    dark = await ui.run_javascript('Quasar.Dark.isActive')
    ui.notify(f'dark mode: {dark}')

ui.button('Get dark mode', on_click=get_dark_mode)
```
We could think about integrating such a function. Maybe `ui.dark_mode.is_dark(): bool`?

But there might be a different approach without the need for calling an async function:
Depending on the user preference, the body tag gets a "body--dark" or "body--light" class. If we could change the first line in jse-theme-dark from
```css
.jse-theme-dark {
```

to something like
```css
body.body--dark .my-json-editor {
```
it might be enough to switch the theme automatically. We do something similar for syntax highlighting in `ui.markdown`:
https://github.com/zauberzeug/nicegui/blob/22d29ec3ff1523b5ca1401e8f8cd30bc8904c153/nicegui/elements/markdown.py#L25-L28
But I'm not sure if this is possible.

**Reply by petergaultney** [reply]
it looks like i can't call `ui.run_javascript` when the page is first rendering, because i haven't yet established a connection to the client. And yet what I want to do is conditionally apply a class to one of the elements I'm rendering. 

It's a bit of a chicken-and-egg situation, obviously. I think I should try your suggestion to hopefully do it statelessly.

An uglier option, if that doesn't work, would be to set an event/timer that fires shortly after page load that causes a request to re-render just the one element with its new class.

**Reply by natankeddem** [reply]
`ui.json_editor` currently requires a bit of work to actually enable dark mode. Just adding the class is not good enough you have to actually import the css as well `jse-theme-dark.css`. You can do this at startup by waiting for the client to connect.
```py
from nicegui import Client, ui

@ui.page('/', dark=True)
async def index(client: Client):
    ui.add_head_html('<link href="static/jse-theme-dark.css" rel="stylesheet">')
    editor = ui.json_editor({"content": {"json": {}}})
    await client.connected()
    dark = await ui.run_javascript('Quasar.Dark.isActive')
    if dark:
        editor.classes("jse-theme-dark")
    ui.notify(f'dark mode: {dark}')

ui.run()
```
